src/frontend/src/lib/utils/qr-code.utils.ts

```diff
@@ -1,7 +1,84 @@
-import type { QrResponse, QrStatus } from '$lib/types/qr-code';
+import {
+	DecodedUrnSchema,
+	URN_NUMERIC_PARAMS,
+	URN_STRING_PARAMS,
+	type DecodedUrn,
+	type QrResponse,
+	type QrStatus
+} from '$lib/types/qr-code';
 import { decodePayment } from '@dfinity/ledger-icrc';
 import { isNullish, nonNullish, type Token } from '@dfinity/utils';
 
+/**
+ * Decodes a URN string into a DecodedUrn object, breaking it down into its components.
+ *
+ * The URN string is expected to follow the pattern defined by the regex:
+ * /^([a-zA-Z]+):([a-zA-Z0-9\-.]+)(@(\d+))?(\/([a-zA-Z]+))?(\?(.*))?$/
+ *
+ * This regex pattern is inspired by multiple sources with each respective URN scheme:
+ * - For IC: https://github.com/dfinity/ICRC/issues/22
+ * - For ETH: https://eips.ethereum.org/EIPS/eip-681
+ * - For BTC: https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki
+ *
+ * @param {string} urn - The URN string to decode.
+ * @returns {DecodedUrn | undefined} The decoded URN object, or undefined if the URN string does not match the expected pattern.
+ */
+export const decodeQrCodeUrn = (urn: string): DecodedUrn | undefined => {
+	const regex = /^([a-zA-Z]+):([a-zA-Z0-9\-.]+)(@(\d+))?(\/([a-zA-Z]+))?(\?(.*))?$/;
+
+	const match = urn.match(regex);
+	if (isNullish(match)) {
+		return undefined;
+	}
+
+	const [_, prefix, destination, , networkId, , functionName, , queryString] = match;
+
+	const processParam = ([key, value]: [string, string]) => {
+		if ((URN_NUMERIC_PARAMS as readonly string[]).includes(key)) {
+			return { [key]: parseFloat(value) };
+		}
+		if ((URN_STRING_PARAMS as readonly string[]).includes(key)) {
+			return { [key]: value };
+		}
+		if (!isNaN(parseFloat(value))) {
+			return { [key]: parseFloat(value) };
+		}
+		return { [key]: value };
+	};
+
+	const parseQueryString = (qs: string): { [key: string]: string | number | undefined } => {
+		try {
+			return [...new URLSearchParams(qs).entries()].reduce(
+				(acc, entry) => ({
+					...acc,
+					...processParam(entry)
+				}),
+				{}
+			);
+		} catch (error) {
+			console.error('Invalid query string:', error);
```

Comment:
I understand this to mean that if the query parameters are invalid, we still decode the rest as valid. If this is correct, shouldn't we rather avoid decoding anything if we are unable to decode the query parameters as well?

I'm not sure if in real life one decodes a QR code, it's surprising or not if not everything is read. So, in a conservative approach, I would spontaneously think about discarding everything.

What do you think? Or maybe I just misunderstood?

Reply:
Yes, sorry, forgot to mention, I addressed it in the last commit

Reply:
Gotcha. Can you either replace the `console.error` with a comment or replace it with a `console.warn` given that we one purpose catch the error and return undefined instead?